pkgs/by-name/fs/fsautocomplete/package.nix

```diff
@@ -36,11 +36,10 @@ buildDotnetModule (finalAttrs: rec {
 
   dotnet-sdk =
     with dotnetCorePackages;
-    combinePackages [
-      sdk_6_0
-      sdk_7_0
-      sdk_8_0
-    ];
+    sdk_8_0
+    // {
+      inherit (sdk_6_0) packages targetPackages;
+    };
```

Comment:
The idea is that we don't actually need more than one SDK, but we do need the targeting packages.  This breaks the dependencies on the insecure packages.

It does feel messy.  I started working on a bigger overhaul of how nuget packages are referenced, but it's not close to being ready.

Reply:
The issue is that the packages themselves being used are also insecure, this feels like just swiping it under the rug as opposed to actually fixing the problem.

This one is gonna be a hard no from me, considering that many .NET CVEs were on SDK packages and not on the compiler/runtime itself.

Reply:
For net 6, it definitely is just sweeping it under the rug, but I still think it's better for things to only depend on the SDKs they use, instead of using `combinePackages` solely to pull in nuget packages.

I'd be happy to mark all net6/7 nuget packages as insecure, but that's a separate problem IMO.

Reply:
Another question, shouldn't we keep the 8.0 packages here (and elsewhere where this is used)?

And couldn't the instances where this uses `combinePackages` be replaced with  the `lib` methods `conmbinePackages` uses instead of creating a full derivation?
It feels wasteful to create a whole combined SDK when we only need the packages and targeting packages from it.

And if possible, I'd like to have the packages be marked as insecure before we start actually using them so this doesn't end up unmarking packages as insecure and people mistakenly think these are safe to use.

Reply:
> Another question, shouldn't we keep the 8.0 packages here (and elsewhere where this is used)?

They just become unused dependencies, so I don't think so.  I'd like to get to the point where you only pull in the nuget packages that are actually needed to target the frameworks/runtimes you need to target, and using the features you need (e.g. optionally ILC).

>  It feels wasteful to create a whole combined SDK when we only need the packages and targeting packages from it.

It shouldn't have to actually evaluate it, since it's just pulling out the passthru packages.

> And if possible, I'd like to have the packages be marked as insecure before we start actually using them so this doesn't end up unmarking packages as insecure and people mistakenly think these are safe to use.

There are so many old nuget packages being used all over the place.  I'm not against doing this, but I'm not sure how far to take it.

Lockfiles are the worst :|

Reply:
> > Another question, shouldn't we keep the 8.0 packages here (and elsewhere where this is used)?
> 
> They just become unused dependencies, so I don't think so.  I'd like to get to the point where you only pull in the nuget packages that are actually needed to target the frameworks/runtimes you need to target, and using the features you need (e.g. optionally ILC).
> 
> >  It feels wasteful to create a whole combined SDK when we only need the packages and targeting packages from it.
> 
> It shouldn't have to actually evaluate it, since it's just pulling out the passthru packages.

Oh, okay, fair enough.

> > And if possible, I'd like to have the packages be marked as insecure before we start actually using them so this doesn't end up unmarking packages as insecure and people mistakenly think these are safe to use.
> 
> There are so many old nuget packages being used all over the place.  I'm not against doing this, but I'm not sure how far to take it.
> 
> Lockfiles are the worst :|

Initially we should mark the EOL SDK packages as insecure, that'd be enough for the .NET 6 and 7 EOL effort.

If anyone fetches a deprecated/obsolete package from nuget on their own I think that's a separate issue that should be addressed but not as urgently as the EOL effort.

Reply:
I'm just having a go at this, by applying `knownVulnerabilities` to the packages as well:

```
$ nix eval -f. fsautocomplete
«error: Package ‘Microsoft.AspNetCore.App.Ref-6.0.36’ in /home/david/src/nixpkgs/pkgs/build-support/dotnet/fetch-nupkg/default.nix:23 is marked as insecure, refusing to evaluate.


Known issues:
 - Dotnet SDK 6.0.428 is EOL, please use 8.0 (LTS) or 9.0 (Current)

You can install it anyway by allowing this package, using the
following methods:

a) To temporarily allow all insecure packages, you can use an environment
   variable for a single invocation of the nix tools:

     $ export NIXPKGS_ALLOW_INSECURE=1

   Note: When using `nix shell`, `nix build`, `nix develop`, etc with a flake,
         then pass `--impure` in order to allow use of environment variables.

b) for `nixos-rebuild` you can add ‘Microsoft.AspNetCore.App.Ref-6.0.36’ to
   `nixpkgs.config.permittedInsecurePackages` in the configuration.nix,
   like so:

     {
       nixpkgs.config.permittedInsecurePackages = [
         "Microsoft.AspNetCore.App.Ref-6.0.36"
       ];
     }

c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
   ‘Microsoft.AspNetCore.App.Ref-6.0.36’ to `permittedInsecurePackages` in
   ~/.config/nixpkgs/config.nix, like so:

     {
       permittedInsecurePackages = [
         "Microsoft.AspNetCore.App.Ref-6.0.36"
       ];
     }
```

It's going to be really painful for anyone who wants to use `permittedInsecurePackages`.

Reply:
Well, there's no way around it unfortunately. I think making it easy for people to use insecure packages isn't an objective of nixpkgs, but worst case they can do something like

```nix
nixpkgs.config.permittedInsecurePackages = lib.map (pkg: pkg.name) pkgs.dotnetCorePackages.sdk_6_0.packages;
```

Reply:
I guess you could make a pseudo-package referenced by the sdk, runtime, nuget packages, and mark that as insecure?

I don't know if it's worth the trouble though.

```
nixpkgs.config.permittedInsecurePackages = lib.map (pkg: pkg.name) pkgs.dotnetCorePackages.sdk_6_0.packages;
```

You'd also need to bring in `targetPackages` which is an attrset of lists.  It's still doable, but cumbersome.

Reply:
> I guess you could make a pseudo-package referenced by the sdk, runtime, nuget packages, and mark that as insecure?

Sounds like too much work for something we plan to remove in the long term, but it would make the deprecation process easier perhaps. If it's easy enough we could do it to make it easier on us to deprecate them as well.

> You'd also need to bring in `targetPackages` which is an attrset of lists. It's still doable, but cumbersome.

Yeah, but I'd say allowing usage of insecure things is a non-objective. If people *really* want to use them, it should be doable but I wouldn't go out of my way to make it easy.